File: webflux-with-mongo-example/src/main/java/com/example/webfluxwithmongoexample/domain/user/service/UserService.java
package com.example.webfluxwithmongoexample.domain.user.service;

import com.example.webfluxwithmongoexample.domain.user.User;
import com.example.webfluxwithmongoexample.domain.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;

    public Mono<UserInfo> save(UserCommand command) {
        User user = new User(command);
        Mono<User> savedUser = userRepository.save(user);
        return savedUser.map(saved -> new UserInfo(saved.getName()));
    }

    public Flux<UserInfo> getAllUsers(UserCommand command) {
        return userRepository.findAll()
                .map(user -> new UserInfo(user.getName()));
    }

    public Mono<UserInfo> findById(String id) {
        return userRepository.findById(id)
                .map(user -> new UserInfo(user.getName()));
    }
}

File: graphql-example/src/main/kotlin/com/example/graphqlexample/service/UserService.kt
package com.example.graphqlexample.service

import com.example.graphqlexample.domain.User
import com.example.graphqlexample.exception.InvalidInputException
import com.example.graphqlexample.exception.NotFoundException
import com.example.graphqlexample.graphql.input.CreateUserInput
import com.example.graphqlexample.repository.UserRepository
import org.springframework.stereotype.Service
import java.time.OffsetDateTime

@Service
class UserService(
    private val userRepository: UserRepository
) {
    fun getUser(userId: Long): User =
        userRepository.findById(userId).orElseThrow { NotFoundException("User not found with id $userId") }

    fun createUser(input: CreateUserInput): User {
        if (input.name.isBlank()) {
            throw InvalidInputException("Name must not be blank")
        }
        if (input.email.isBlank()) {
            throw InvalidInputException("Email must not be blank")
        }
        if (userRepository.existsByEmail(input.email)) {
            throw InvalidInputException("Email already exists")
        }

        val user = User(
            name = input.name.trim(),
            email = input.email.trim(),
            createdAt = OffsetDateTime.now()
        )
        return userRepository.save(user)
    }

    fun search(keyword: String): List<User> =
        userRepository.findByNameContainingIgnoreCase(keyword)
}
